### PR TITLE
bump rust stable to 1.72.1

### PR DESCRIPTION
**Pull request by t-nelson**
#### Problem

rust stable 1.72.0 is [apparently not always stable](https://blog.rust-lang.org/2023/09/19/Rust-1.72.1.html)

#### Summary of Changes

* bump rust stable to 1.72.1
* bump rust nightly to 2023-09-20
* fix new lints

### CODE REVIEW

**Comment by t-nelson** [comment]
working on docker images now

**Comment by t-nelson** [comment]
how'd we get away without `cargo hack` being installed in the previous nightly image?

**Review by ilya-bobyr — DISMISSED** [review]

File: ci/docker-rust-nightly/Dockerfile [hunk 1]
```diff
@@ -9,6 +9,7 @@ RUN set -x \
  && rustc --version \
  && cargo --version \
  && cargo install grcov \
+ && cargo install cargo-hack \
```
Comment (yihau) [inline ci/docker-rust-nightly/Dockerfile:12]: I think we have installed this one in the stable image 🤔 
https://github.com/solana-labs/solana/blob/2839d51d5ee1a5d379b00b89062c6ff142b163f0/ci/docker-rust/Dockerfile#L37

let me check

**Comment by ilya-bobyr** [comment]
> how'd we get away without `cargo hack` being installed in the previous nightly image?

It seems to be used just in a few locations (ignoring comments):
```
❯ git grep -w hack
ci/docker:rust/Dockerfile:37: && cargo install cargo-hack \
ci/test-checks.sh:29:   $cargoNightly hack --version --verbose
scripts/check-dev-context-only-utils.sh:148:  _ cargo "+${rust_nightly}" hack check --bins
scripts/check-dev-context-only-utils.sh:151:  _ cargo "+${rust_nightly}" hack check --all-targets
```

I wonder if `ci/test-checks.sh` and `scripts/check-dev-context-only-utils.sh` are never run in the nightly Docker container.

File: ci/docker-rust-nightly/Dockerfile (hunk 1 above)
Comment (ilya-bobyr) [reply ci/docker-rust-nightly/Dockerfile:12]: Yes, it is in the stable `Dockerfile`:
```
❯ git grep -w hack
ci/docker-rust/Dockerfile:37: && cargo install cargo-hack \
[...]
```

Comment (t-nelson) [reply ci/docker-rust-nightly/Dockerfile:12]: i just figured it out.  accidentally built stable from wrong shell which had 1.16 checked out :upside_down_face:

Comment (yihau) [reply ci/docker-rust-nightly/Dockerfile:12]: fwiw, I'm building docker image with a new method for installing node

**Comment by yihau** [comment]
just retried the pipeline. let's see how it goes

**Comment by t-nelson** [comment]
ok i think it'll go this time

**Review by ilya-bobyr — APPROVED** [review]

**Pull request merged**